File: server/src/main/java/xyz/chunkstories/server/net/ClientsManager.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.server.net;

import xyz.chunkstories.api.player.Player;
import xyz.chunkstories.net.http.SimpleWebRequest;
import xyz.chunkstories.server.DedicatedServer;
import xyz.chunkstories.server.DedicatedServerOptions;
import xyz.chunkstories.server.player.ServerPlayer;
import xyz.chunkstories.util.VersionInfo;

import java.util.Iterator;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public abstract class ClientsManager {

	protected final DedicatedServer server;
	protected final ServerPacketsProcessorImplementation packetsProcessor;

	protected String hostname;
	protected String externalIP;

	protected int maxClients;

	protected Set<ClientConnection> clients = ConcurrentHashMap.newKeySet();

	protected Map<String, ServerPlayer> players = new ConcurrentHashMap<>();
	protected Map<Long, ServerPlayer> playersByUUID = new ConcurrentHashMap<>();

	public ClientsManager(DedicatedServer server) {
		this.server = server;
		this.packetsProcessor = new ServerPacketsProcessorImplementation(server);

		externalIP = new SimpleWebRequest("https://chunkstories.xyz/api/sayMyName.php?ip=1").waitForResult();
		hostname = new SimpleWebRequest("https://chunkstories.xyz/api/sayMyName.php?host=1").waitForResult();

		this.maxClients = server.getServerConfig().getIntValue(DedicatedServerOptions.INSTANCE.getMaxUsers());
	}

	public DedicatedServer getServer() {
		return server;
	}

	public abstract boolean open();

	public void flushAll() {
		for (ClientConnection client : clients)
			client.flush();
	}

	public boolean close() {
		Iterator<ClientConnection> clientsIterator = getAllConnectedClients();
		while (clientsIterator.hasNext()) {
			ClientConnection client = clientsIterator.next();
			// Remove the client first to avoid concurrent mod exception
			clientsIterator.remove();

			client.disconnect("Server is closing.");
		}

		return true;
	}

	public int getMaxClients() {
		return maxClients;
	}

	public Iterator<ClientConnection> getAllConnectedClients() {
		return clients.iterator();
	}

	public String getIP() {
		return externalIP;
	}

	public String getHostname() {
		return hostname;
	}

	public ServerPacketsProcessorImplementation getPacketsProcessor() {
		return this.packetsProcessor;
	}

	public int getPlayersNumber() {
		return players.size();
	}

	public Player getPlayerByName(String playerName) {
		return players.get(playerName);
	}

	public Set<Player> getPlayers() {
		return new LinkedHashSet(players.values());
		/*return new IterableIterator<Player>() {
			Iterator<ServerPlayer> authClients = players.values().iterator();

			@Override
			public boolean hasNext() {
				return authClients.hasNext();
			}

			@Override
			public ServerPlayer next() {
				return authClients.next();
			}
		};*/
	}

	public void sendServerInfo(ClientConnection clientConnection) {
		clientConnection.sendTextMessage("info/name:" + getServer().getServerConfig().getValue(DedicatedServerOptions.INSTANCE.getServerName()));
		clientConnection.sendTextMessage("info/motd:" + getServer().getServerConfig().getValue(DedicatedServerOptions.INSTANCE.getServerDescription()));
		clientConnection.sendTextMessage("info/connected:" + getPlayersNumber() + ":" + getMaxClients());
		clientConnection.sendTextMessage("info/version:" + VersionInfo.version);
		clientConnection.sendTextMessage("info/mods:" + getServer().getModsProvider().getModsString());
		clientConnection.sendTextMessage("info/done");

		// We flush because since the potential player isn't registered, the automatic
		// flush at world ticking doesn't apply to them
		clientConnection.flush();
	}

	/**
	 * Used by ClientConnection after a successfull login procedure
	 */
	void registerPlayer(ClientConnection clientConnection) {
		ServerPlayer player = clientConnection.getPlayer();
		this.players.put(player.getName(), player);
		this.playersByUUID.put(player.getUuid(), player);
	}

	/**
	 * Used by ClientConnection during the close() method
	 */
	void removeClient(ClientConnection clientConnection) {
		this.clients.remove(clientConnection);

		ServerPlayer player = clientConnection.getPlayer();
		if (player != null) {
			this.players.remove(player.getName());
			this.playersByUUID.remove(player.getUuid());
		}
	}

}

File: server/src/main/java/xyz/chunkstories/server/net/ClientsManager.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.server.net

import xyz.chunkstories.api.player.Player
import xyz.chunkstories.net.http.SimpleWebRequest
import xyz.chunkstories.server.DedicatedServer
import xyz.chunkstories.server.DedicatedServerOptions
import xyz.chunkstories.server.player.ServerPlayer
import xyz.chunkstories.util.VersionInfo
import java.util.*
import java.util.concurrent.ConcurrentHashMap

abstract class ClientsManager(val server: DedicatedServer) {
    val packetsProcessor: ServerPacketsProcessorImplementation

    var hostname: String
        protected set
    var ip: String
        protected set

    var maxClients: Int = 0
        protected set

    protected var clients: MutableSet<ClientConnection> = ConcurrentHashMap.newKeySet()

    protected var playersMap: MutableMap<String, ServerPlayer> = ConcurrentHashMap()
    protected var playersByUUID: MutableMap<Long, ServerPlayer> = ConcurrentHashMap()

    val allConnectedClients: MutableIterator<ClientConnection>
        get() = clients.iterator()

    val playersNumber: Int
        get() = playersMap.size

    init {
        this.packetsProcessor = ServerPacketsProcessorImplementation(server)

        ip = SimpleWebRequest("https://chunkstories.xyz/api/sayMyName.php?ip=1").waitForResult()
        hostname = SimpleWebRequest("https://chunkstories.xyz/api/sayMyName.php?host=1").waitForResult()

        this.maxClients = server.serverConfig.getIntValue(DedicatedServerOptions.maxUsers)
    }

    abstract fun open(): Boolean

    fun flushAll() {
        for (client in clients)
            client.flush()
    }

    open fun close(): Boolean {
        val clientsIterator = allConnectedClients
        while (clientsIterator.hasNext()) {
            val client = clientsIterator.next()
            // Remove the client first to avoid concurrent mod exception
            clientsIterator.remove()

            client.disconnect("Server is closing.")
        }

        return true
    }

    fun getPlayerByName(playerName: String): Player? {
        return playersMap[playerName]
    }

    val players: Set<Player>
        get() {
            return LinkedHashSet(playersMap.values)
        }

    fun sendServerInfo(clientConnection: ClientConnection) {
        clientConnection.sendTextMessage("info/name:" + server.serverConfig.getValue(DedicatedServerOptions.serverName))
        clientConnection.sendTextMessage("info/motd:" + server.serverConfig.getValue(DedicatedServerOptions.serverDescription))
        clientConnection.sendTextMessage("info/connected:$playersNumber:$maxClients")
        clientConnection.sendTextMessage("info/version:" + VersionInfo.versionJson.version)
        clientConnection.sendTextMessage("info/mods:" + server.modsProvider.modsString)
        clientConnection.sendTextMessage("info/done")

        // We flush because since the potential player isn't registered, the automatic
        // flush at world ticking doesn't apply to them
        clientConnection.flush()
    }

    /**
     * Used by ClientConnection after a successfull login procedure
     */
    internal fun registerPlayer(clientConnection: ClientConnection) {
        val player = clientConnection.player
        this.playersMap[player!!.name] = player
        this.playersByUUID[player.uuid] = player
    }

    /**
     * Used by ClientConnection during the close() method
     */
    internal fun removeClient(clientConnection: ClientConnection) {
        this.clients.remove(clientConnection)

        val player = clientConnection.player
        if (player != null) {
            this.playersMap.remove(player.name)
            this.playersByUUID.remove(player.uuid)
        }
    }

}